src/experimental/tools/videoThumbnailLoader/thumbnail_loader.ts

```diff
@@ -0,0 +1,245 @@
+  private readonly _videoElement: HTMLVideoElement;
+
+  private _player: Player;
+  private _currentJob?: IJob;
+
+  constructor(videoElement: HTMLVideoElement,
+              player: Player) {
+    this._videoElement = videoElement;
+    this._player = player;
+  }
+
+  /**
+   * Set time of thumbnail video media element :
+   * - Remove buffer when too much buffered data
+   * - Search for thumbnail track element to display
+   * - Load data
+   * - Append data
+   * Resolves when time is set.
+   * @param {number} time
+   * @returns {Promise}
+   */
+  setTime(time: number): Promise<unknown> {
+    return this._setTime(time, this._videoElement);
+  }
+
+  /**
+   * Dispose thumbnail loader.
+   * @returns {void}
+   */
+  dispose(): void {
+    disposeSourceBuffer(this._videoElement);
+    this._currentJob?.stop();
+  }
+
+  private _setTime = (time: number,
+                      videoElement: HTMLVideoElement
+  ): Promise<unknown> => {
+
+    for (let i = 0; i < videoElement.buffered.length; i++) {
+      if (videoElement.buffered.start(i) <= time &&
+          videoElement.buffered.end(i) >= time) {
+        this._videoElement.currentTime = time;
+        log.debug("VTL: Thumbnail already loaded.");
+        return PPromise.resolve(time);
+      }
+    }
+
+    const manifest = this._player.getManifest();
+    if (manifest === null) {
+      return PPromise.reject(new VideoThumbnailLoaderError("NO_MANIFEST",
+                                                           "No manifest available."));
+    }
+    const contentInfos = getContentInfos(time, manifest);
+    if (contentInfos === null) {
+      return PPromise.reject(new VideoThumbnailLoaderError("NO_TRACK",
+                                                           "Couldn't find track for this time."));
+    }
+    const initURL = contentInfos.representation.index.getInitSegment()?.mediaURL ?? "";
+    if (initURL === "") {
+      return PPromise.reject(new VideoThumbnailLoaderError("NO_INIT_DATA", "No init data for track."));
+    }
+    const segment = getNeededSegment(contentInfos,
+                                     time);
+    if (segment === undefined) {
+      return PPromise.reject(new VideoThumbnailLoaderError("NO_THUMBNAILS", "Couldn't find thumbnail."));
+    }
+
+    log.debug("VTL: Found thumbnail for time", time, segment);
+
+    if (this._currentJob === undefined) {
+      return this.startJob(contentInfos, time, segment);
+    }
+
+    if (this._currentJob.contentInfos.representation.getMimeTypeString() !==
+        contentInfos.representation.getMimeTypeString()) {
+      this._currentJob.stop();
+      return this.startJob(contentInfos, time, segment);
+    }
+    if (this._currentJob.segment.time !== segment.time ||
+        this._currentJob.segment.duration !== segment.duration ||
+        this._currentJob.segment.mediaURL !== segment.mediaURL) {
+      this._currentJob.stop();
+      return this.startJob(contentInfos, time, segment);
+    }
```

Comment:
Code organization from 140 to 154 makes not much sense:
  - we always call `startJob` with the same arguments
  - there are two `if` with exactly the same result

Why not just a single `if` for stopping the currentJob (we could use `this._currentJob?.stop()) and a single call to `startJob`?

Reply:
👍